src/pybind/mgr/cephadm/module.py

```diff
@@ -2481,6 +2551,136 @@ def upgrade_stop(self):
         self.event.set()
         return trivial_result('Stopped upgrade to %s' % target_name)
 
+    def save_spec(self, spec: ServiceSpec) -> None:
+        """
+        Attempts to save a ServiceSpec.
+
+        There are two ways to manipulate the service_specs stored in the mon_store
+        1) Using a global definition (i.e. `ceph orch apply -i <specs_file.yaml>`)
+            This will usually contain multiple definitions of services and daemons.
+
+        2) Using the CLI for specific service types (i.e. `ceph orch apply rgw 3 --realm foo --zone bar`)
+
+         Raises `OrchestratorError` if an entry with the same named identifier already exist to prevent overwrites from two different sources.
+        """
+        store = PersistentStoreDict(self, f'service_spec/{spec.service_type}')
+        name: str = spec.name or spec.service_type
+        try:
+            _ = store[name]
+            raise orchestrator.OrchestratorError(f"Specification for {name} already exists. "
+                                                 "Please review your existing specs with "
+                                                 "`ceph orch servicespecs ls` and try again.")
+        except IndexError:
+            self.log.info(f"New spec found. Saving <{name}> to the store.")
+            store[name] = spec.to_json()
+
+    def list_specs(self) -> orchestrator.Completion:
+        """
+        Loads all entries from the service_spec mon_store root.
+        """
+        specs = list()
+        for _, spec in self.service_spec_store.items():
+            specs.append('---')
+            specs.append(yaml.dump(spec))
+        return trivial_result(specs)
+
+    def clear_all_specs(self) -> orchestrator.Completion:
+        """
+        Clears the service_spec root in the mon_store (mgr/cephadm/service_specs/)
+        """
+        self.service_spec_store.clear()
+        return trivial_result(f"Mon store for {self.service_spec_store.prefix} cleared")
+
+    def apply_service_config(self, spec_document: str) -> orchestrator.Completion:
+        """
+        Parse a multi document yaml file (represented in a inbuf object)
+        and loads it with it's respective ServiceSpec to validate the
+        initial input.
+        If no errors are raised `save_spec` is called.
+        """
+        content: Iterator[Any] = yaml.load_all(spec_document)
+        # Load all specs from a multi document yaml file.
+        for spec in content:
+            # load ServiceSpec once to validate
+            spec_o = ServiceSpec.from_json(spec)
```

Comment:
Note that this will save specs until an exception is raised. This may result in partial success. Either the user is advised to fix their input and re-apply the service spec _or_ we're parsing all specs before saving a single one.

Reply:
i'd suggesting parsing/validating them all, and then saving them.

Reply:
ack